src/features/generic_events/aggregate/index.js

```diff
@@ -0,0 +1,114 @@
+
+    this.referrerUrl = undefined
+    this.currentEvents = []
+
+    this.events = []
+    this.overflow = []
+
+    this.#agentRuntime = getRuntime(this.agentIdentifier)
+
+    if (isBrowserScope && document.referrer) this.referrerUrl = cleanURL(document.referrer)
+
+    this.waitForFlags(['ins']).then(([ins]) => {
+      if (!ins) {
+        this.blocked = true
+        deregisterDrain(this.agentIdentifier, this.featureName)
+        return
+      }
+
+      // handle page actions and other generic events here
+      this.harvestScheduler = new HarvestScheduler('ins', { onFinished: (...args) => this.onHarvestFinished(...args) }, this)
+      this.harvestScheduler.harvest.on('ins', (...args) => this.onHarvestStarted(...args))
+      // this.harvestScheduler.startTimer(this.harvestTimeSeconds, 0)
+
+      this.drain()
+    })
+  }
+
+  onHarvestStarted (options) {
+    const { userAttributes, atts } = getInfo(this.agentIdentifier)
+    const harvestEvents = this.overflow.length ? this.overflow.splice(0, Infinity) : this.events.splice(0, Infinity)
+    var payload = ({
+      qs: {
+        ua: userAttributes,
+        at: atts
+      },
+      body: {
+        ins: harvestEvents
+      }
+    })
+
+    if (options.retry) {
+      this.currentEvents = harvestEvents
+    }
+
+    return payload
+  }
+
+  onHarvestFinished (result) {
+    if (result && result.sent && result.retry && this.currentEvents.length) {
+      this.events = this.currentEvents.concat(this.events)
+      this.currentEvents = []
+    }
+  }
+
+  // WARNING: Insights times are in seconds. EXCEPT timestamp, which is in ms.
+  addEvent (obj = {}) {
+    if (!obj || !Object.keys(obj).length) return
+    if (!obj.eventType) {
+      warn('Invalid object passed to generic event aggregate. Missing "eventType".')
+      return
+    }
+
+    for (let key in obj) {
+      let val = obj[key]
+      if (key === 'timestamp') val = this.#agentRuntime.timeKeeper.correctAbsoluteTimestamp(val)
+      obj[key] = (val && typeof val === 'object' ? stringify(val) : val)
+    }
+
+    const eventAttributes = {
+      /** Agent-level custom attributes */
+      ...(getInfo(this.agentIdentifier).jsAttributes || {}),
+      /** Common attributes shared on all generic events */
+      referrerUrl: this.referrerUrl,
+      currentUrl: cleanURL('' + location),
+      pageUrl: cleanURL(getRuntime(this.agentIdentifier).origin),
+      /** Event-specific attributes take precedence over everything else */
+      ...obj
+    }
+
+    /** should have been provided by reporting feature -- but falls back to now if not */
+    eventAttributes.timestamp ??= this.#agentRuntime.timeKeeper.convertRelativeTimestamp(now())
+
+    this.events.push(eventAttributes)
+
+    // check if we've reached the harvest limit...
+    if (this.events.length >= this.eventsPerHarvest) {
+      this.overflow = [...this.overflow, ...this.events.splice(0, Infinity)]
+      this.harvestScheduler.runHarvest()
```

Comment:
We should put in a ticket to centralize the harvest buffer mechanics so that we are not re-inventing the wheel in every feature around max harvest size, etc.

Reply:
yeah, needs a middle man to buffer instead of feature by feature so we can check, but agreed